root: Change the default value of the runtime_cxxmodules option

jmcarcell opened the pull request:
Since ROOT 6.20 the default (https://github.com/root-project/root/pull/4452) is to have `runtime_cxxmodules` ON and we have it OFF by default with no way of configuring it, so I propose to follow the default value and set it to ON. I have found that in our stack having this set to off causes some slow downs in some IO operations (https://github.com/key4hep/key4hep-spack/issues/743). The PCH option was removed in https://github.com/root-project/root/commit/898171f5a4619ac07245d70b7f8943da067516ba so I propose not to set it.

jmcarcell commented:
The failure in CI looks very weird to me. It's Gaudi, and it seems the tests are being run while building? Since what I'm proposing here is to follow the defaults, I'm quite sure Gaudi can be built with these changes.

marcmengel reviewed (approved):
Looks good.

alalazo reviewed (changes requested):
Can you check this failure: https://gitlab.spack.io/spack/spack/-/jobs/16755256 ?

marcmengel commented:
That is a very strange build error; it looks like the C++ compiler is somehow getting the  preprocessor, input, rather than the preprocessor output,  and thus getting things like #define and #pragma on its input, and complaining about them as syntax errors?!?

gartung reviewed (changes requested):
Runtime modules should not be enabled by default as this breaks the Gaudi tests.

gartung commented:
It would be better to change both `pch` and `runtime_modules` build options to `define_from_variant`.

gartung commented:
The corollary is that you need to define a Variant for each as well. Variant allows you to set a default value.

jmcarcell committed:
root: set runtime_cxxmodules to True for newer versions

following the default in ROOT. Don't set pch since that was removed in
898171f5a4619ac07245d70b7f8943da067516ba. When runtime_cxxmodules is set to
False in recent versions, a big file is created that is read and some IO
operations are slowed down, see
https://github.com/key4hep/key4hep-spack/issues/743.

haampie commented:
Hi @jmcarcell, this package-related pull request can be migrated from `spack/spack` to `spack/spack-packages` using the [**migrate-pkg-prs utility**](https://github.com/spack/migrate-package-prs).

We encourage authors to run the migration script themselves to preserve author attribution.

You can migrate all your open pull requests at once, following the steps in the [documentation](https://github.com/spack/migrate-package-prs?tab=readme-ov-file#usage-instructions).

---

*This comment was posted automatically.*
